In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

2.

In [2]:
df.shape

(891, 15)

Observatioans are simply individual data points that are being measured and collected about a given variable. Each observation corresponds to a single occurrence of something being measured. In this dataset, observations (rows) represent individual passengers.

Variables represent characteristics or values that can be measured or counted for each observation. Each variable shows different features and attributes of observations. In this dataset, variables (columns) represent characteristics such as survived, pclass, and age.

3.

In [4]:
df.survived.describe()

count    891.000000
mean       0.383838
std        0.486592
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: survived, dtype: float64

In [6]:
df['pclass'].value_counts()

pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [7]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Basic information about the dataset
print("Basic Info:")
df.info()

# Summary statistics for numerical columns
print("\nSummary Statistics for Numerical Columns:")
print(df.describe())

# Value counts for categorical columns
print("\nValue Counts for Categorical Columns:")
for col in df.columns:
    if df[col].dtype == 'object' or df[col].nunique() < 10:
        print(f"\nColumn: {col}")
        print(df[col].value_counts())

Basic Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB

Summary Statistics for Numerical Columns:
         survived      pclass         age       sibsp 

In [8]:
df.shape

(891, 15)

In [9]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

4.

df.shape() gives the total number of rows. The "count" column in df.describe() gives the number of non-missing values in each column. Columns with missing data will show a lower count than the total number of rows. This is the reason why the column with the variable "age" has 714 instead of 891. Also, df.shape reports all columns in the dataset, but df.describe() by default only includes numeric columns, so it may analyze fewer columns.

5.

An attribute is a value or property associated with an object (like a DataFrame) and is used to store information about the object. Attributes do not require parenthese () when accessed. They are like variables that store the state or charateristics of the object. 
For example, df.shape is an attribute that stores the dimensions of the DataFrame. It simply provides the information without any additional processing or computation when you access it. It does not perform any action.

Methods are functions that performs an action or computation on the object and typically require parentheses () to be executed. They require parenthese because they may also include arguments to modify how the method behaves.
For example, df. describe() is a method that performs a operation on the DataFrame. It calculates summary statistics (mean, count, min etc.) for the numeric columns.

In summary, attributes are properties or values that hold information about the object (no computation), while methods are functions that perform an action or calculation on the object.

Last step: Summary of my ChatBot session with ChatGPT

Summary of Interaction for Assignment
In this interaction, I worked on analyzing the Titanic dataset available at https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv using Python and the Pandas library. The key topics covered include:

Loading and Summarizing the Dataset:

I was guided on how to load the Titanic dataset using Pandas' read_csv() function.
I was shown how to determine the number of rows and columns using the attribute df.shape.
Understanding Observations and Variables:

I learned the meaning of observations (rows) and variables (columns) in the context of the dataset. An observation represents data for one passenger, and a variable represents attributes like age, gender, fare, etc.
Exploring Data Columns:

A detailed explanation was provided for each column in the dataset (e.g., survived, pclass, sex, etc.) along with insights derived from these columns, such as the survival rate and class distribution.
Summarizing Data with df.describe() and df['column'].value_counts():

I learned how to summarize numerical columns using df.describe(), which provides statistics like the mean, standard deviation, minimum, and maximum values.
For categorical data, I used df['column'].value_counts() to understand the distribution of unique values within the dataset (e.g., how many passengers embarked from each port or how many survived).
Simplest Dataset Summarization:

I was introduced to the simplest methods of summarizing a dataset using df.info() to check data types, missing values, and overall structure, alongside df.describe() and value counts for an efficient overview of both numerical and categorical data.
Difference Between Attributes and Methods:

I explored the difference between attributes (e.g., df.shape, which provides stored information) and methods (e.g., df.describe(), which performs an action to compute statistics).
This interaction has helped me better understand how to work with datasets in Pandas, summarizing both numerical and categorical data effectively.

https://chatgpt.com/share/09eae353-5512-4d41-8041-75cfbcbc5429

Post-lecture HW Below

6.

Count: The number of valid data points or values that are present in the dataset for each variable. It tells you the number of all the non-missing data points.

Mean: It is the average value of the data points in a column and it's calculated by summing up all the values and dividing by the count of those values.

Standard Deviation (std): A measure of how dispersed the data points are in relation to the mean. It tells you how much the values typically deviate from the mean.

Min: The smallest value in the dataset for a given column.

25% (First Quartile): The value under which 25% of data points lie below that value.

50% (Median): The middle value of the dataset when it is arranged in increasing order. Half the data points are below this value, and the other half are above it.

75% (Third Quartile): The value below which 75% of the data falls. It divides the top quarter of the dataset from the rest.

Max: The largest value in the dataset for a given column.

7.

a) df.dropna() might be preferred over using del df['col'] when only a few rows in the dataset have missing values. It is more efficient to drop the rows with missing values rather than deleting entire columns, which could result in the loss of important information. 
For example, if there are 500 rows and 5 columns in the dataset where only 3 rows have missing values in various columns, using df.dropna() to remove these 3 rows would preserve 497 rows of data. It's efficient because you're only losing a small portion of the dataset. If the del df['col'] is used, and some columns had missing values in those 3 rows, deleting the entire column would result in a larger data loss, especially if the column has valuable information.

b) Using del df['col'] might be preferred over using df.dropna() when a particular column has a high percentage of missing values or contains mostly unusable data. Therefore, it's better to drop the entire column rather than removing rows, which could lead to losing a significant amount of valid data from other columns.
For example, if there's a dataset with 500 rows and 5 columns, but one column has 90% missing values, using df.dropna() would result in removing almost all the rows. Instead, using del df['col'] to remove that specific column would preserve the rest of the dataset while eliminating a mostly unusable column.

c) Applying del df['col'] first helps focus on the essential columns. Once unnecessary columns with missing values are removed, df.dropna() will be applied more effectively, keeping the maximum number of valid rows in the essential columns. Deleting irrelevant columns first with del df['col'] ensures that when I drop rows with missing values (df.dropna()), I am only focusing on the important data, preventing unnecessary data loss and preserving more useful information for my analysis.

d)

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

before:

In [5]:
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

after:

In [23]:
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
Index: 889 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     889 non-null    int64  
 1   pclass       889 non-null    int64  
 2   sex          889 non-null    object 
 3   sibsp        889 non-null    int64  
 4   parch        889 non-null    int64  
 5   fare         889 non-null    float64
 6   embarked     889 non-null    object 
 7   class        889 non-null    object 
 8   who          889 non-null    object 
 9   adult_male   889 non-null    bool   
 10  embark_town  889 non-null    object 
 11  alive        889 non-null    object 
 12  alone        889 non-null    bool   
dtypes: bool(2), float64(1), int64(4), object(6)
memory usage: 85.1+ KB


survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

In [12]:
del df['deck']

In [14]:
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [15]:
df_clean = df.dropna()

In [16]:
df.dropna(inplace=True)

In [17]:
df_clean = df.dropna(subset=['embarked', 'embark_town'])

In [18]:
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

With the help from ChatGPT, I removed all the missing values from the dataset. First, I used del df['col'] to remove the entire columns that have a significant amount of missing or irrelevant data ('age', 'deck'). It was more efficient to delete the whole columns because they had too many missing values. Then I used df.dropna() to remove rows that have missing values in important columns ('embarked', 'embark town'). Now in the dataset, there are no missing values.

8. a)

df.groupby("col1") basically groups the DataFrame by the values in column "col1". Each unique value in "col1" will form a separate group. 
["col2"].describe(): After grouping, this applies the describe() method to column "col2" within each group, providing summary statistics such as count, mean, std, min, 25%, 50%, 75%, and max for "col2" within each group.

In [28]:
df.groupby("adult_male")["alive"].describe()

count unique  top freq
adult_male                       
False        352      2  yes  252
True         537      2   no  449

In [29]:
df.groupby("sex")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  312.0  0.740385  0.439128  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

b)

One of the key differences is the scope of analysis. df.describe() provides and overall view of missing data for each column across the entire Dataframe, while df.groupby("col1")["col2"].describe() provides a view of missing data for "col2" within each group defined by "col1". This can show how missing values in "col2" are distributed across different groups in "col1". 

Another difference is in the count calculation. The count in df.describe relfects the number of non-missing values for each column across all rows. The count in df.groupby("col1")["col2"].describe() reflects the number of non-missing values for "col2" within each group of "col1". This means that if particular group defined by "col1" has fewer non-missing values in "col2", the count will be lower for that group.

df.describe(): missing values in different columns will affect the count for each column independently.
df.groupby("col1")["col2"].describe(): missing values in "col2" will be assessed separately for each group. Therefore, the count might show variability based on how many non-missing values are present within each group of "col1".

c)

A.

In [1]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

NameError: name 'pd' is not defined

For me, it is way easier to work in ChatBot session to fix the errors because it tells me the answer straightforward. However, if I search it up on google then I would usually have to look for a useful website that will tell me the answer, which takes an extra step.

B.

In [3]:
import pandas as pd
url = "titanics.csv"
df = pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [4]:
import pandas as pd
url = "https://raw.gihubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

URLError: <urlopen error [Errno -2] Name or service not known>

For this one, I would say neither because both ChatGPT and Google aren't telling me the reason explicitly. If I were to choose at least one, then I would choose ChatGPT because it shows me an example of an code right away so it would be helpful to find out why the error occurs.

C.

In [5]:
DF.groupby("col1")["col2"].describe()

NameError: name 'DF' is not defined

ChatGPT is much useful in this one because it tells me the solution right away and tells me that DF and df are treated as two different variables.

D.

In [6]:
pd.read_csv(url

SyntaxError: incomplete input (4098100527.py, line 1)

ChatGTP was easier to work with because it tells me multiple possible reasons why the error occured and it helps me resolve the problem faster.

E.

In [16]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

In [19]:
df.group_by("col1")["col2"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

In [17]:
df.groupby("col1")["col2"].describle()

KeyError: 'col1'

ChatGPT was easier again because it shows me examples and codes so I can find out very quickly. It also tells me straightforward that I've mistakenly written "group_by" instead of "groupby".

F.

In [20]:
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

In [21]:
df.groupby("sex")["Age"].describe()

KeyError: 'Column not found: Age'

ChatGPT tells me that I should use 'sex' instead of 'Sex'. Same for 'age' as well. This chatbot makes it very easy for me to resolve the problems.

G.

In [22]:
df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

In [23]:
df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

ChatGPT tells me right away that "This usually happens if the string is not enclosed in quotes or if there is a syntax error." so it is very helpful. I tried searching up on Google and it took longer to find the resolution.

9.

Yes

Summary of my ChatBot session with ChatGPT

Summary of Interaction
Objective: The goal was to perform initial simple summary analyses on the Titanic dataset and resolve common errors encountered in the process.

Steps and Issues Addressed:

Understanding df.describe():

Explained the summary statistics provided by the describe() method, including count, mean, std, min, 25%, 50%, 75%, and max.
Handling Missing Values:

Discussed methods to handle missing values, including df.dropna() to remove rows with missing data and df.fillna() to fill missing values.
Code Execution Issues:

Encountered and resolved issues related to:
NameError: Explained how this occurs when a variable is used before it is defined and the importance of correct variable names.
URLError: Addressed issues with loading datasets from URLs and provided solutions for local file loading.
AttributeError: Corrected the use of groupby instead of group_by.
KeyError: Ensured the correct use of column names, including case sensitivity and spaces.
SyntaxError: Addressed incomplete input issues and the importance of proper syntax.
NameError for column names: Clarified the need to enclose column names in quotes.
Demonstration with Titanic Dataset:

Provided examples for loading the Titanic dataset and performing groupby operations with correct syntax and column names.
Key Takeaways:

Ensure column names are correctly referenced and enclosed in quotes.
Use the correct methods and syntax to handle DataFrame operations.
Verify that the dataset is correctly loaded and accessible.

https://chatgpt.com/share/3db72f53-c88d-4481-b78b-8d5ad9902404